core/src/main/java/org/elasticsearch/common/path/PathTrie.java

```diff
@@ -233,4 +233,56 @@ public T retrieve(String path, Map<String, String> params) {
         }
         return root.retrieve(strings, index, params);
     }
+
+    /**
+     * Splits up the url path up by '/' and is aware of
+     * index name expressions that appear between '<' and '>'.
+     */
+    String[] splitPath(final String path) {
+        if (path == null || path.length() == 0) {
+            return Strings.EMPTY_ARRAY;
+        }
+        int count = 1;
+        boolean splitAllowed = true;
+        for (int i = 0; i < path.length(); i++) {
+            final char currentC = path.charAt(i);
+            if ('<' == currentC) {
+                splitAllowed = false;
+            } else if (currentC == '>') {
+                splitAllowed = true;
+            } else if (splitAllowed && currentC == separator) {
+                count++;
+            }
+        }
+        final String[] result = new String[count];
+        final StringBuilder builder = new StringBuilder();
+
+        int res = 0;
+        splitAllowed = true;
+        for (int i = 0; i < path.length(); i++) {
+            final char currentC = path.charAt(i);
+            if ('<' == currentC) {
+                splitAllowed = false;
+            } else if (currentC == '>') {
+                splitAllowed = true;
+            } else  if (splitAllowed && currentC == separator) {
+                if (builder.length() > 0) {
+                    result[res++] = builder.toString();
+                    builder.setLength(0);
+                }
+                continue;
+            }
+            builder.append(currentC);
+        }
+        if (builder.length() > 0) {
+            result[res++] = builder.toString();
+        }
+        if (res != count) {
+            // we have empty strings, copy over to a new array
+            String[] result1 = new String[res];
+            System.arraycopy(result, 0, result1, 0, res);
+            return result1;
+        }
```

Comment:
more of a question than a comment: seems to me that using an array here complicates the code a bit. I guess that is for performance reasons? Otherwise we could use a list and iterate only once, and also save us the arraycopy at the end?